src/logic/equiv/set.lean

```diff
@@ -512,6 +512,37 @@ begin
   by_cases hi : p i; simp [hi]
 end
 
+/-- An equation in the target type gives an equivalence between the corresponding preimages. -/
+@[simps]
+def preimage_congr {α β : Type*} (f : α → β)
+  {c c' : β} (h : c = c') : f ⁻¹' {c} ≃ f ⁻¹' {c'} :=
+{ to_fun := λ x, ⟨x.1, by simpa [h] using x.2⟩,
+  inv_fun := λ x, ⟨x.1, by simpa [h] using x.2⟩,
+  left_inv := λ x, by simp,
+  right_inv := λ x, by simp, }
```

Comment:
Golfs to
```suggestion
equiv.set_congr $ h ▸ rfl
```

Reply:
I like the golfs, but I still don't see how to prove the main result without having these in hand. :-)